src/ngraph/runtime/executable.hpp

```diff
@@ -73,6 +73,36 @@ class ngraph::runtime::Executable
     ///    Saved stream may be read with Backend::load
     virtual void save(std::ostream& output_stream);
 
+    /// \brief Create an input Tensor
+    /// \param input_index The index position in the input Parameter vector. This would be the same
+    /// order of Parameters passed into the inputs in the call() method.
+    /// \returns A Tensor
+    virtual std::shared_ptr<runtime::Tensor> create_input_tensor(size_t input_index);
+
+    /// \brief Create an output Tensor
+    /// \param output_index The index position in the output Result vector. This would be the same
+    /// order of Results passed into the outputs in the call() method.
+    /// \returns A Tensor
+    virtual std::shared_ptr<runtime::Tensor> create_output_tensor(size_t output_index);
+
+    /// \brief Create a vector of input Tensors
+    /// \param input_index The index position in the input Parameter vector. This would be the same
+    /// order of Parameters passed into the inputs in the call() method.
+    /// \param pipeline_depth The number of stages in the input pipeline. For double-buffered input
+    /// you would specify pipeline_depth=2
+    /// \returns A vector of Tensors, one for each stage of the pipeline
+    virtual std::vector<std::shared_ptr<runtime::Tensor>> create_input_tensor(size_t input_index,
+                                                             size_t pipeline_depth);
+
+    /// \brief Create a vector of output Tensors
+    /// \param output_index The index position in the output Result vector. This would be the same
+    /// order of Results passed into the outputs in the call() method.
+    /// \param pipeline_depth The number of stages in the output pipeline. For double-buffered output
+    /// you would specify pipeline_depth=2
+    /// \returns A vector of Tensors, one for each stage of the pipeline
+    virtual std::vector<std::shared_ptr<runtime::Tensor>> create_output_tensor(size_t output_index,
+                                                              size_t pipeline_depth);
+
```

Comment:
General questions: 
Going forward, will this be the recommended way to create input and output tensors?

Will the executable call() be able to take in inputs/outputs created by the backend->create_tensor() PI? 
https://github.com/NervanaSystems/ngraph/blob/master/src/ngraph/runtime/backend.hpp#L68

Right now, theoretically it is possible to pass in the same input tensors to different executables (if the shape and size match) will that be still a possibility, say if we store the tensor?

Reply:
@shresthamalik These new methods would be preferable because they tell us more.
Your question re tying together is a good one. That should still be ok, but we'll need to ensure it.
Let's chat in brief meeting next week.

Reply:
I would like to deprecate the Backend methods for creating tensors but that is a big issue and left for later.

Reply:
There are a number of optimizations we can do if we can guarantee the tensors which are passed in to multiple functions/executables are handled similarly. 

For example, if tensor0 (layout NHWC) is used by executableA and executableB, if both executables have the same boundary layout transformation (NHWC -> NCHW) then we can fuse the reshape into the DMA-write pipeline and skip the reshape on device for both executables A and B. But if executable A does the NHWC->NCHW conversion, but executable B does not, then we can't make the optimization.

We will handle this decision making in the backend, but the bridge teams should be aware of it so as to ensure optimal layout scheduling for the executable where possible.